More robust error handling

jfly opened the pull request:
More robust error handling

I really like this plugin, but I think the error messages and behavior
when you don't have the right plugins installed (or the wrong versions
of those binaries provided by plugins) could be improved. That's what
I've done here:

- Prevent creating the cached env file if there are any errors.
- Raise an error if we depend on a plugin that isn't installed. (This
  fixes https://github.com/asdf-community/asdf-direnv/issues/49)
- Raise an error if we depend on a version of a tool that isn't
  installed.

Before
======

Setup:

    $ cat .tool-versions
    maven 3.8.4
    $ cat .envrc
    use asdf

Try loading direnv, note that we're missing the `maven` plugin:

    $ direnv reload
    direnv: loading ~/tmp/bugging/.envrc
    direnv: using asdf
    direnv: Creating env file /home/jeremy/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-3896840882-2383759277
    No such plugin: maven
    direnv: loading ~/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-3896840882-2383759277
    direnv: using asdf direnv 2.30.3
    direnv: using asdf maven 3.8.4
    direnv: export ~PATH

If you look carefully, you'll see a "No such plugin: maven" in there, but we
went on to do other things. For example, note how that env cache file got
created, which means even if we add the maven plugin, we're going to have to do
something to bust the cache to fix this:

    $ cat /home/jeremy/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-3896840882-2383759277
    log_status using asdf direnv 2.30.3
    PATH_add /home/jeremy/.asdf/installs/direnv/2.30.3/bin
    watch_file /home/jeremy/.tool-versions
    log_status using asdf maven 3.8.4
    PATH_add
    watch_file /home/jeremy/tmp/bugging/.tool-versions

Install the maven plugin. Now how we reuse that env cache file:

    $ asdf plugin add maven
    $ cd ..; cd -
    direnv: unloading
    ~/tmp/bugging
    direnv: loading ~/tmp/bugging/.envrc
    direnv: using asdf
    direnv: loading ~/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-3896840882-2383759277
    direnv: using asdf direnv 2.30.3
    direnv: using asdf maven 3.8.4
    direnv: export ~PATH

Now blow the cache:

    $ touch .envrc
    direnv: loading ~/tmp/bugging/.envrc
    direnv: using asdf
    direnv: Creating env file /home/jeremy/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-472850511-2383759277
    direnv: loading ~/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-472850511-2383759277
    direnv: using asdf direnv 2.30.3
    direnv: using asdf maven 3.8.4
    direnv: export ~PATH

Ok, no errors. But I know I don't have maven 3.8.4 installed. What's going to happen here?

    $ which mvn
    mvn not found
    $ mvn --version
    zsh: command not found: mvn
    $ echo $PATH
    /home/jeremy/.asdf/installs/maven/3.8.4/bin:...
    $ ls -alh /home/jeremy/.asdf/installs/maven/3.8.4/bin
    ls: cannot access '/home/jeremy/.asdf/installs/maven/3.8.4/bin': No such file or directory

Ooh so you can see that even when maven 3.8.4 isn't installed, we end up with a
nonexistent folder on the `PATH` identifying where that plugin will eventually
be installed.

Install the correct version of maven:

    $ asdf install maven 3.8.4

And we immediately start finding `mvn`:

    $ which mvn
    /home/jeremy/.asdf/installs/maven/3.8.4/bin/mvn

After
=====

Setup:

    $ cat .tool-versions
    maven 3.8.4
    $ cat .envrc
    use asdf

Try loading direnv, note that we're missing the `maven` plugin:

    $ direnv reload
    direnv: loading ...
    direnv: using asdf
    direnv: Creating env file ...
    direnv: asdf plugin not installed: maven
    direnv: Error generating asdf cached envrc

Add the plugin and try again (note: we don't have the right version of
`maven` installed yet):

    $ asdf plugin add maven
    $ cd ..; cd -
    direnv: loading ~/tmp/bugging/.envrc
    direnv: using asdf
    direnv: Creating env file ...
    direnv: maven 3.8.4 not installed. Run 'asdf install' and try again.
    direnv: Error generating asdf cached envrc

Install the missing version and try again:

    $ asdf install
    ...
    $ cd ..; cd -
    direnv: loading ~/tmp/bugging/.envrc
    direnv: using asdf
    direnv: Creating env file ...
    direnv: loading ~/.asdf/installs/direnv/2.30.3/env/1560215600-3608870685-1742556474-2383759277
    direnv: using asdf direnv 2.30.3
    direnv: using asdf maven 3.8.4
    direnv: export ~PATH
    $ which mvn
    /home/jeremy/.asdf/installs/maven/3.8.4/bin/mvn

vic commented:
Hey @jfly, this is awesome, the code looks good to me :) Could. you fix the lint and formatting errors ? Thank you a lot !

jfly commented:
@vic could you check again? I believe I've fixed the lint and formatting errors.

I also discovered a bug: I wasn't actually preventing creation of the cached env file... I thought I was, but I didn't realize that `direnv reload` forces things to start over even if there's already a cache file.

I fixed this by instead writing to a `${hash}.new` file before `mv`-ing the file to `${hash}`. That way if we crash midway through generating the file, we aren't left with a broken hashed env file.

vic commented:
nice, thank you @jfly

jfly commented:
Thank you for the quick response!
